go/vt/wrangler/stream_migrater.go

```
@@ -0,0 +1,604 @@
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"text/template"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"golang.org/x/net/context"
+	"vitess.io/vitess/go/mysql"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/binlog/binlogplayer"
+	"vitess.io/vitess/go/vt/concurrency"
+	"vitess.io/vitess/go/vt/key"
+	binlogdatapb "vitess.io/vitess/go/vt/proto/binlogdata"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/throttler"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/vterrors"
+	"vitess.io/vitess/go/vt/vttablet/tabletmanager/vreplication"
+)
+
+type streamMigrater struct {
+	mi *migrater
+}
+
+type vrStream struct {
+	id       uint32
+	workflow string
+	bls      *binlogdatapb.BinlogSource
+	pos      mysql.Position
+}
+
+func (sm *streamMigrater) stopStreams(ctx context.Context) ([]*vrStream, error) {
+	if sm.mi.migrationType == binlogdatapb.MigrationType_TABLES {
+		// Source streams should be stopped only for shard migrations.
+		return nil, nil
+	}
+	streams, err := sm.readSourceStreams(ctx)
+	if err != nil {
+		return nil, err
+	}
+	streams, err = sm.stopSourceStreams(ctx, streams)
+	if err != nil {
+		return nil, err
+	}
+	positions, err := sm.syncSourceStreams(ctx, streams)
+	if err != nil {
+		return nil, err
+	}
+	return sm.verifyStreamPositions(ctx, streams, positions)
+}
+
+func (sm *streamMigrater) readSourceStreams(ctx context.Context) (map[string][]*vrStream, error) {
+	streams := make(map[string][]*vrStream)
+	var mu sync.Mutex
+	err := sm.mi.forAllSources(func(source *miSource) error {
+		stoppedStreams, err := sm.readTabletStreams(ctx, source.master, "state = 'Stopped'")
+		if err != nil {
+			return err
+		}
+		if len(stoppedStreams) != 0 {
+			return fmt.Errorf("cannot migrate until all strems are running: %s", source.si.ShardName())
+		}
+		tabletStreams, err := sm.readTabletStreams(ctx, source.master, "")
+		if err != nil {
+			return err
+		}
+		if len(tabletStreams) == 0 {
+			return nil
```

Should return a proper error message.

---

This is legitimate and the most common case. This means that there are no other vreplications running. So, stream migrater has no work to do. Added clarification.

---

Oh this comment was supposed to be 2 lines higher, where we are doing `return err`.